safe_transaction_service/events/services/queue_service.py

```diff
@@ -0,0 +1,272 @@
+        by sending the Channel.OpenOK RPC reply, the on_channel_open method
+        will be invoked.
+
+        """
+        logger.info("Opening a new channel")
+        self._connection.channel(on_open_callback=self.on_channel_open)
+
+    def on_channel_open(self, channel: Channel):
+        """This method is invoked by pika when the channel has been opened.
+        The channel object is passed in so we can make use of it.
+
+        :param pika.channel.Channel channel: The channel object
+
+        """
+        logger.info("Channel with number %i opened", channel.channel_number)
+        self._channel = channel
+        self._channel.add_on_close_callback(self.on_channel_closed)
+        self.setup_exchange()
+
+    def on_channel_closed(self, channel: Channel, reason: Exception):
+        """Invoked by pika when RabbitMQ unexpectedly closes the channel.
+        Channels are usually closed if you attempt to do something that
+        violates the protocol.
+        In this method we retry to open a new channel with rabbitMQ if the connection is still open.
+
+        :param Channel channel: The closed channel
+        :param Exception reason: why the channel was closed
+
+        """
+        logger.warning("Channel %i was closed: %s", channel.channel_number, reason)
+        self._channel = None
+        if self._connection and self._connection.is_open:
+            # If channel was closed and connection is still active we try to reopen the channel
+            logger.error(
+                "Connection is opened retry to open channel in 5 seconds: %s",
+                self._connection_parameters.host,
+                reason,
+            )
+            self._connection.ioloop.call_later(5, self.open_channel())
+
+    def setup_exchange(self):
+        """Setup the exchange on RabbitMQ by invoking the Exchange.Declare RPC
+        command. When it is complete, the on_exchange_declareok method will
+        be invoked by pika.
+
+        """
+        logger.info("Declaring exchange %s", self.EXCHANGE_NAME)
+
+        self._channel.exchange_declare(
+            exchange=self.EXCHANGE_NAME,
+            exchange_type=ExchangeType.fanout,
+            durable=True,
+            callback=self.on_exchange_declareok,
+        )
+
+    def on_exchange_declareok(self, _unused_frame):
+        """Invoked by pika when RabbitMQ has finished the Exchange.Declare RPC
+        command.
+        Send unsent messages that cannot be sent as due connection errors.
+
+        :param pika.Frame.Method unused_frame: Exchange.DeclareOk response frame
+        """
+
+        logger.info("Exchange declared: %s", self.EXCHANGE_NAME)
+        self.send_unsent_events()
+
+
+class SyncQueueService(QueueService):
+    """
+    Synchronous connection with test purpose as we cannot test using gevent connection
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.connect()
+
+    def connect(self):
```

Comment:
I think you should return the connection when connecting, even if you are assigning locally